File: generated_code/GPT4/Java/020_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    /** Inserts a key into the set (no duplicates). */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }
        Node z = root, p = null;
        while (z != null) {
            p = z;
            if (key < z.key)
                z = z.left;
            else if (key > z.key)
                z = z.right;
            else {
                splay(z);
                return; // key already exists, no duplicates
            }
        }
        Node n = new Node(key);
        n.parent = p;
        if (key < p.key)
            p.left = n;
        else
            p.right = n;
        splay(n);
    }

    /** Deletes a key from the set (if exists). */
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null)
            return;
        splay(node);
        // Remove node
        if (node.left == null) {
            replace(node, node.right);
        } else if (node.right == null) {
            replace(node, node.left);
        } else {
            // Find max in left subtree
            Node max = subtreeMax(node.left);
            if (max.parent != node) {
                replace(max, max.left);
                max.left = node.left;
                if (max.left != null)
                    max.left.parent = max;
            }
            replace(node, max);
            max.right = node.right;
            if (max.right != null)
                max.right.parent = max;
        }
        // Remove all references
        node.left = node.right = node.parent = null;
    }

    /** Searches for a key. Splays the node to root if found. Returns the key, or null if not found. */
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // --------- Internal Helper Methods ---------

    private Node findNode(int key) {
        Node z = root;
        while (z != null) {
            if (key < z.key)
                z = z.left;
            else if (key > z.key)
                z = z.right;
            else
                return z;
        }
        return null;
    }

    // Splaying operation: moves x to the root
    private void splay(Node x) {
        if (x == null)
            return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig
                if (x.parent.left == x)
                    rightRotate(x.parent);
                else
                    leftRotate(x.parent);
            } else if (x.parent.left == x && x.parent.parent.left == x.parent) {
                // Zig-Zig
                rightRotate(x.parent.parent);
                rightRotate(x.parent);
            } else if (x.parent.right == x && x.parent.parent.right == x.parent) {
                // Zig-Zig
                leftRotate(x.parent.parent);
                leftRotate(x.parent);
            } else if (x.parent.left == x && x.parent.parent.right == x.parent) {
                // Zig-Zag
                rightRotate(x.parent);
                leftRotate(x.parent);
            } else {
                // Zig-Zag
                leftRotate(x.parent);
                rightRotate(x.parent);
            }
        }
        root = x;
    }

    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null)
            return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    private void rightRotate(Node x) {
        Node y = x.left;
        if (y == null)
            return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Replace node u with node v in the tree
    private void replace(Node u, Node v) {
        if (u.parent == null)
            root = v;
        else if (u == u.parent.left)
            u.parent.left = v;
        else
            u.parent.right = v;
        if (v != null)
            v.parent = u.parent;
    }

    // Finds the maximum node in a subtree
    private Node subtreeMax(Node x) {
        while (x.right != null)
            x = x.right;
        return x;
    }

    // --------- Optional: Utility Methods ---------

    /** Inorder traversal as a string, for debugging. */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        inorder(root, sb);
        return sb.toString();
    }

    private void inorder(Node node, StringBuilder sb) {
        if (node == null)
            return;
        inorder(node.left, sb);
        sb.append(node.key).append(' ');
        inorder(node.right, sb);
    }

    // --------- Main method for demonstration ---------
    public static void main(String[] args) {
        SplayTree st = new SplayTree();
        st.insert(10);
        st.insert(20);
        st.insert(30);
        st.insert(40);
        st.insert(50);
        System.out.println("Tree (inorder): " + st);

        System.out.println("Search 30: " + st.search(30)); // Should splay 30 to root
        System.out.println("Tree (inorder): " + st);

        st.delete(20);
        System.out.println("After deleting 20: " + st);

        System.out.println("Search 60: " + st.search(60)); // Not found
        System.out.println("Tree (inorder): " + st);
    }
}

File: generated_code/GPT4/Python/020_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ['key', 'left', 'right', 'parent']
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # Right rotation
    def _rotate_right(self, x):
        y = x.left
        if not y:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.right = x
        x.parent = y

    # Left rotation
    def _rotate_left(self, x):
        y = x.right
        if not y:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    # Splaying operation: bring x to root
    def _splay(self, x):
        if not x:
            return
        while x.parent:
            if not x.parent.parent:
                # Zig case
                if x == x.parent.left:
                    self._rotate_right(x.parent)
                else:
                    self._rotate_left(x.parent)
            else:
                p = x.parent
                g = p.parent
                if x == p.left and p == g.left:
                    # Zig-zig case (left-left)
                    self._rotate_right(g)
                    self._rotate_right(p)
                elif x == p.right and p == g.right:
                    # Zig-zig case (right-right)
                    self._rotate_left(g)
                    self._rotate_left(p)
                elif x == p.right and p == g.left:
                    # Zig-zag case (left-right)
                    self._rotate_left(p)
                    self._rotate_right(g)
                elif x == p.left and p == g.right:
                    # Zig-zag case (right-left)
                    self._rotate_right(p)
                    self._rotate_left(g)

    # Search and splay
    def search(self, key):
        x = self.root
        last = None
        while x:
            last = x
            if key == x.key:
                self._splay(x)
                return True
            elif key < x.key:
                x = x.left
            else:
                x = x.right
        # Not found, splay the last accessed node (parent of None)
        if last:
            self._splay(last)
        return False

    # Insert key
    def insert(self, key):
        if not self.root:
            self.root = self.Node(key)
            return

        x = self.root
        p = None
        while x:
            p = x
            if key == x.key:
                self._splay(x)
                return  # Already present, do nothing
            elif key < x.key:
                x = x.left
            else:
                x = x.right

        new_node = self.Node(key, parent=p)
        if key < p.key:
            p.left = new_node
        else:
            p.right = new_node
        self._splay(new_node)

    # Delete key
    def delete(self, key):
        if not self.root:
            return

        if not self.search(key):
            return  # Key not found

        # Now root is the node to delete
        node = self.root
        if not node.left:
            self.root = node.right
            if self.root:
                self.root.parent = None
        elif not node.right:
            self.root = node.left
            if self.root:
                self.root.parent = None
        else:
            # Find the maximum node in left subtree
            left_sub = node.left
            left_sub.parent = None
            max_left = left_sub
            while max_left.right:
                max_left = max_left.right
            # Splay max_left to root of left subtree
            self._splay(max_left)
            max_left.right = node.right
            if node.right:
                node.right.parent = max_left
            self.root = max_left
            self.root.parent = None

    # (Optional) For debugging: inorder traversal
    def inorder(self):
        res = []
        def _inorder(node):
            if node:
                _inorder(node.left)
                res.append(node.key)
                _inorder(node.right)
        _inorder(self.root)
        return res